src/reconcile.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for calculating required changes based on docs directory and navigation table."""
```

Comment:
How about using Python standard library [`difflib`](https://docs.python.org/3/library/difflib.html)? `difflib` is not just for diff texts, you can use `difflib.SequenceMatcher` to generate opcodes.

Reply:
Thanks for the link the the library, I hadn't used that before. It will be helpful for providing a human readable difference between the old and new form of the documentation which will be helpful to include in the actions and action reports. In terms of using it for the logic in this module, it is somewhat limited because the way the documentation is stored in the repo versus the server is quite different:

- On the server, each documentation page is stored as a discourse post, equivalent to a URL of the form `https://<discourse host>/t/<slug>/<topic id>. There is no folder or nested documentation indicated in these links. That information comes from the navigation table on a separate index post.
- In the repo, each file is stored in a directory based on the page hierarchy. For example, if a documentation file is stored in a folder, it indicates nested documentation. The file and directory structure, file names and information from an index file is used to calculate the navigation table.

These different ways of representing the documentation have to be understood and reconciled in this module. It could be possible to build adapters to enable the use of this library, although that will probably result in more complexity than could be saved. The opcodes also seem to be more designed based on a git diff, the create, update, noop and delete are more web op codes as, for example, rather than individual lines, entire documentation pages are updated on the server.